src/otx/algo/diffusion/huggingface_model.py

```diff
@@ -0,0 +1,315 @@
+        self.pipe.text_encoder.eval()
+        self.pipe.vae.eval()
+
+    def _create_model(self) -> nn.Module:
+        return UNetWrapper(self.pipe.unet)
+
+    def forward_for_tracing(
+        self,
+        sample: torch.Tensor,
+        timestep: torch.Tensor,
+        encoder_hidden_states: torch.Tensor,
+    ) -> torch.Tensor | dict[str, torch.Tensor]:
+        """Model forward function used for the model tracing during model exportation."""
+        return self.model(sample, timestep, encoder_hidden_states)
+
+    def _customize_inputs(
+        self,
+        entity: DiffusionBatchDataEntity,
+        pad_size_divisor: int = 32,
+        pad_value: int = 0,
+    ) -> dict[str, Any]:
+        if isinstance(entity.images, list):
+            entity.images, entity.imgs_info = stack_batch(
+                entity.images,
+                entity.imgs_info,
+                pad_size_divisor=pad_size_divisor,
+                pad_value=pad_value,
+            )
+        input_ids = self.pipe.tokenizer(
+            entity.captions,
+            max_length=self.pipe.tokenizer.model_max_length,
+            padding="max_length",
+            truncation=True,
+            return_tensors="pt",
+        ).input_ids.to(self.device)
+        latents = self.pipe.vae.encode(entity.images).latent_dist.sample()
+        latents = latents * self.pipe.vae.config.scaling_factor
+
+        # Sample noise that we'll add to the latents
+        noise = torch.randn_like(latents)
+        # Sample a random timestep for each image
+        timesteps = torch.randint(
+            0,
+            self.pipe.scheduler.config.num_train_timesteps,
+            (latents.shape[0],),
+            device=latents.device,
+        )
+        timesteps = timesteps.long()
+
+        # Add noise to the latents according to the noise magnitude at each timestep
+        # (this is the forward diffusion process)
+        noisy_latents = self.pipe.scheduler.add_noise(latents, noise, timesteps)
+
+        # Get the text embedding for conditioning
+        encoder_hidden_states = self.pipe.text_encoder(input_ids, return_dict=False)[0]
+
+        if self.pipe.scheduler.config.prediction_type == "epsilon":
+            target = noise
+        elif self.pipe.scheduler.config.prediction_type == "v_prediction":
+            target = self.pipe.scheduler.get_velocity(latents, noise, timesteps)
+        else:
+            error_msg = f"Invalid prediction type: {self.pipe.scheduler.config.prediction_type}"
+            raise ValueError(error_msg)
+
+        return {
+            "sample": noisy_latents,
+            "timestep": timesteps,
+            "encoder_hidden_states": encoder_hidden_states,
+            "target_noise": target,
+        }
+
+    def _customize_outputs(
+        self,
+        outputs: torch.Tensor,
+        inputs: DiffusionBatchDataEntity,
+    ) -> DiffusionBatchPredEntity | OTXBatchLossEntity:
+        return OTXBatchLossEntity(loss=outputs)
```

Comment:
It seems better to move into `core/model/diffusion` because it can be commonly used with other diffucision models to be added. What do you think?

Reply:
Fixed in [4a82d0d](https://github.com/openvinotoolkit/training_extensions/pull/3875/commits/4a82d0d2aa2f5b16803f10ac9f2d1725dd542682)